src/contracts/orderbook/NiftyswapOrderbook.sol

```diff
@@ -0,0 +1,422 @@
+            }
+        }
+
+        // Transfer currency
+        TransferHelper.safeTransferFrom(order.currency, tokenReceiver, currencyReceiver, remainingCost);
+
+        // Transfer token
+        address tokenContract = order.tokenContract;
+        if (_tokenIsERC1155(tokenContract)) {
+            IERC1155(tokenContract).safeTransferFrom(currencyReceiver, tokenReceiver, order.tokenId, quantity, "");
+        } else {
+            IERC721(tokenContract).transferFrom(currencyReceiver, tokenReceiver, order.tokenId);
+        }
+    }
+
+    /**
+     * Cancels a listing.
+     * @param listingId The ID of the listing.
+     */
+    function cancelListing(bytes32 listingId) external {
+        Order storage listing = orders[listingId];
+        if (listing.creator != msg.sender) {
+            revert InvalidListingId(listingId);
+        }
+        address tokenContract = listing.tokenContract;
+
+        // Refund some gas
+        delete orders[listingId];
+
+        emit ListingCancelled(listingId, tokenContract);
+    }
+
+    /**
+     * Cancels an offer.
+     * @param offerId The ID of the offer.
+     */
+    function cancelOffer(bytes32 offerId) external {
+        Order storage offer = orders[offerId];
+        if (offer.creator != msg.sender) {
+            revert InvalidOfferId(offerId);
+        }
+        address tokenContract = offer.tokenContract;
+
+        // Refund some gas
+        delete orders[offerId];
+
+        emit OfferCancelled(offerId, tokenContract);
+    }
+
+    /**
+     * Deterministically create the orderId for the given order.
+     * @param order The order.
+     * @return orderId The ID of the order.
+     */
+    function hashOrder(Order memory order) public pure returns (bytes32 orderId) {
+        return keccak256(
+            abi.encodePacked(
+                order.creator,
+                order.tokenContract,
+                order.tokenId,
+                order.quantity,
+                order.currency,
+                order.pricePerToken,
+                order.expiry
+            )
+        );
+    }
+
+    /**
+     * Gets an order.
+     * @param orderId The ID of the listing or offer.
+     * @return order The order.
+     */
+    function getOrder(bytes32 orderId) external view returns (Order memory order) {
+        return orders[orderId];
+    }
+
+    /**
+     * Checks if orders are valid.
+     * @param orderIds The IDs of the orders.
+     * @return valid The validities of the orders.
+     * @notice An order is valid if it is active, has not expired and tokens (currency for offers, tokens for listings) are transferrable.
+     */
+    function isOrderValid(bytes32[] memory orderIds) external view returns (bool[] memory valid) {
+        valid = new bool[](orderIds.length);
+        for (uint256 i; i < orderIds.length; i++) {
+            Order memory order = orders[orderIds[i]];
+            valid[i] = order.creator != address(0) && !_isExpired(order)
```

Comment:
Shouldn't be checking `_hasApprovedCurrency` if the order is an offer?

Reply:
Added